x/ecocredit/server/genesis.go

```diff
@@ -1,32 +1,177 @@
 package server
 
 import (
+	"bytes"
+	"encoding/binary"
 	"encoding/json"
 
 	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/pkg/errors"
+	"github.com/regen-network/regen-ledger/orm"
+	"github.com/regen-network/regen-ledger/types/math"
 	abci "github.com/tendermint/tendermint/abci/types"
 
 	"github.com/regen-network/regen-ledger/types"
 	"github.com/regen-network/regen-ledger/x/ecocredit"
 )
 
-
 // InitGenesis performs genesis initialization for the ecocredit module. It
 // returns no validator updates.
 func (s serverImpl) InitGenesis(ctx types.Context, cdc codec.JSONCodec, data json.RawMessage) ([]abci.ValidatorUpdate, error) {
 	var genesisState ecocredit.GenesisState
 	cdc.MustUnmarshalJSON(data, &genesisState)
+
+	s.idSeq.InitVal(ctx, genesisState.IdSeq)
+
 	s.paramSpace.SetParamSet(ctx.Context, &genesisState.Params)
+
+	if err := orm.ImportTableData(ctx, s.classInfoTable, genesisState.ClassInfos, 0); err != nil {
+		return nil, errors.Wrap(err, "class-infos")
+	}
+
+	if err := orm.ImportTableData(ctx, s.batchInfoTable, genesisState.BatchInfos, 0); err != nil {
+		return nil, errors.Wrap(err, "batch-infos")
+	}
+
+	store := ctx.KVStore(s.storeKey)
+	for _, balance := range genesisState.TradableBalances {
+		addr, err := sdk.AccAddressFromBech32(balance.Address)
+		if err != nil {
+			return nil, err
+		}
+		tradable, err := math.ParseNonNegativeDecimal(balance.Balance)
+		if err != nil {
+			return nil, err
+		}
+		key := TradableBalanceKey(addr, batchDenomT(balance.BatchDenom))
+		setDecimal(store, key, tradable)
+	}
+
+	for _, rBalance := range genesisState.RetiredBalances {
+		addr, err := sdk.AccAddressFromBech32(rBalance.Address)
+		if err != nil {
+			return nil, err
+		}
+		retired, err := math.ParseNonNegativeDecimal(rBalance.Balance)
+		if err != nil {
+			return nil, err
+		}
+		key := RetiredBalanceKey(addr, batchDenomT(rBalance.BatchDenom))
+		setDecimal(store, key, retired)
+	}
+
+	for _, tSupply := range genesisState.TradableSupplies {
+		key := TradableSupplyKey(batchDenomT(tSupply.BatchDenom))
+		supply, err := math.ParseNonNegativeDecimal(tSupply.Supply)
+		if err != nil {
+			return nil, err
+		}
+		setDecimal(store, key, supply)
+	}
+
+	for _, rSupply := range genesisState.RetiredSupplies {
+		key := RetiredSupplyKey(batchDenomT(rSupply.BatchDenom))
+		supply, err := math.ParseNonNegativeDecimal(rSupply.Supply)
+		if err != nil {
+			return nil, err
+		}
+		setDecimal(store, key, supply)
+	}
+
+	for _, precision := range genesisState.Precisions {
+		key := MaxDecimalPlacesKey(batchDenomT(precision.BatchDenom))
+		setUInt32(store, key, precision.MaxDecimalPlaces)
+	}
+
 	return []abci.ValidatorUpdate{}, nil
 }
 
+// ExportGenesis will dump the contents of the keeper into a serializable GenesisState.
```

Comment:
strictly speaking, there's no keeper in x/ecocredit so maybe we should upgrade the language here